source/simulator/helper_functions.cc

```diff
@@ -1236,15 +1236,17 @@ namespace aspect
   bool
   Simulator<dim>::stokes_matrix_depends_on_solution() const
   {
-    // currently, the only coefficient that really appears on the
-    // left hand side of the Stokes equation is the viscosity. note
-    // that our implementation of compressible materials makes sure
-    // that the density does not appear on the LHS.
-    // if melt transport is included in the simulation, we have an
+    // Currently, the only coefficient that really appears on the
+    // left hand side of the Stokes equation is the viscosity and possibly
+    // the density in the case of the implicit reference density profile
+    // approximation.
+    // If melt transport is included in the simulation, we have an
     // additional equation with more coefficients on the left hand
     // side.
 
     return (material_model->get_model_dependence().viscosity != MaterialModel::NonlinearDependence::none)
+           || (parameters.formulation_mass_conservation ==
+               Parameters<dim>::Formulation::MassConservation::implicit_reference_density_profile)
```

Comment:
Hm, interesting catch. Did we just forget this before?

Is it easy to understand why the implicit reference density affects the viscosity? The density doesn't appear in the lhs of the Stokes equation, or does it?

Reply:
In this formulation the density appears on the left-hand side. We simply move the explicit reference density profile term to the left-hand side, and hope that the resulting asymmetry of the matrix does not matter too much for the preconditioner (see source/simulator/assembly.cc:316-360). @tjhei created a table in our 2017 paper that shows that the change increases the number of iterations compared to the explicit case with only one nonlinear iteration, but it needs less total iterations than iterating out the explicit version (at least for that example case). 
I think it did not matter that we forgot to add this condition, because we rebuild the matrix in essentially any case anyway, but this makes the condition explicit.

Reply:
Ah, this makes sense now. I had somehow thought that `implicit_reference_density` has to do with the material model, not the formulation. Should have read the namespace name as well! :-)